### PR TITLE
cmake: check for python(\d)\.(\d+) when building boost

### PR DESCRIPTION
**Pull request by tchaikov**
now that python3.10 has two digits as its minor version, we should be
able to match it.

this change fixes the build with python3.10. without this change, we'd
have

error: wrong library name 'python310' in the --with-<library> option.

when building with python3.10

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [x] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by cbodley — APPROVED** [review]

File: cmake/modules/BuildBoost.cmake [hunk 1]
```diff
@@ -67,10 +67,10 @@ function(do_build_boost root_dir version)
 
   set(boost_with_libs)
   foreach(c ${Boost_BUILD_COMPONENTS})
-    if(c MATCHES "^python([0-9])\$")
```
Comment (tchaikov) [inline cmake/modules/BuildBoost.cmake:70]: probably we would never see python10. and according to Guido, python4 might never come to the world in the foreseeable future. so i dropped this change.

**Comment by tchaikov** [comment]
changelog

- do not match `python(\d)+`. as we are not likely to have `python10` in future.

**Comment by zaitcev** [comment]
I confirm that the above commit allows to build Ceph on Fedora 35 with Python 3.10.

[2108/2108] Linking CXX executable bin/dbstore-tests
[zaitcev@lembas build]$ rpm -q python3-devel
python3-devel-3.10.0-1.fc35.x86_64
[zaitcev@lembas build]$ (cd ../ && git log) | head 
commit 66b1e8cecdbabd76a131dc239f6f41d5374fbd32
Author: Kefu Chai <kchai@redhat.com>
Date:   Fri Nov 19 00:18:13 2021 +0800

    cmake: check for python(\d)\.(\d+) when building boost

**Comment by bnavigator** [comment]
Can we get this backported to ceph 16 please? It blocks the openSUSE Tumbleweed update to Python 3.10:

https://build.opensuse.org/package/show/openSUSE:Factory:Staging:A/ceph

**Comment by tchaikov** [comment]
@bnavigator see #46365

**Pull request merged**